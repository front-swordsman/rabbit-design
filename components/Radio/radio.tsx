import * as React from 'react'
import cx from 'classnames'
import * as warning from 'warning'
import { RadioProps, RadioItemProps } from 'types/radio'
import { noop } from '../utils/tool'
import RadioItem from './radioItem'

const { useCallback } = React

const prefixCls = 'snake-radio'

const Radio: React.SFC<RadioProps> & {
  item: React.ForwardRefExoticComponent<RadioItemProps & React.RefAttributes<{}>>
} = ({
  options = [],
  onChange = noop,
  size = 'normal',
  disabled = false,
  shape = 'default',
  buttonStyle = 'outline',
  ...rest
}) => {
  const { className, value } = rest

  const getClassName = useCallback(() => {
    return cx(
      prefixCls,
      {
        [`${prefixCls}-${size}`]: size,
        [`${prefixCls}-disabled`]: disabled,
        [`${prefixCls}-${shape}`]: shape,
        [`${prefixCls}-${buttonStyle}`]: buttonStyle
      },
      className
    )
  }, [size, disabled, shape, buttonStyle, className])

  const handleChange = useCallback((value: string | number) => {
    onChange(value)
  }, [])

  warning(!(value === undefined), 'Radio', 'should have `value` prop')

  return (
    <div className={getClassName()}>
      {options.map(p => {
        const checked = p.value === value
        return (
          <RadioItem
            key={p.value}
            checked={checked}
            disabled={disabled || p.disabled}
            onChange={() => handleChange(p.value)}
          >
            {p.label}
          </RadioItem>
        )
      })}
    </div>
  )
}

Radio.item = RadioItem

export default Radio
