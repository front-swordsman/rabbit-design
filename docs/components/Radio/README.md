---
imports:
  import RadioDemoCode from '!raw-loader!./demo.tsx';
  import ButtonDemoCode from '!raw-loader!./buttonDemo.tsx';
  import RadioDemo from './demo.tsx';
  import Block from '../../layout/block';
  import RadioButtonDemo from './buttonDemo.tsx';
---

# Radio 单选框

用于在多个备选项中选中单个状态

## 简单示例

:::demo
<Block des="常用的 radio" code={RadioDemoCode}>
  <RadioDemo />
</Block>
:::

## 填充示例

:::demo
<Block des="按钮式的 radio" code={ButtonDemoCode}>
  <RadioButtonDemo />
</Block>
:::

## Api

### Radio

| 参数     | 说明              | 类型   | 可选值 | 默认值 |
| -------- | ----------------- | ------ | ------ | ------ |
| options | 数据源 | Data[] | -- | []
| onChange | 触发事件的回调 | func(value: string or number) | -- | noop
| size | 尺寸 | `small`, `normal`, `large` | -- | `normal`
| value | 选中的值 | string, number | 否 | --
| disabled | 不可用 | boolean | -- | `false`
| className | 类名 | string | -- | --
| shape | 形状, 设置为 `button` 则为按钮类型 | `default`, `button` | -- | `default`
| buttonStyle | `button` 类型时候的样式, 分别是填充和描边, 需设置 `shape` 才会生效 | `outline`, `solid` | -- | `outline`

### RadioItem

| 参数     | 说明              | 类型   | 可选值 | 默认值 |
| -------- | ----------------- | ------ | ------ | ------ |
| autoFocus | 是否自动聚焦 | boolean | -- | `false`
| checked | 是否选中 | boolean | -- | `false`
| disabled | 是否不可用 | boolean | -- | `false`
| onChange | 触发事件的回调 | function(checked: boolean): void | -- | noop
| className | 类名 | string | -- | --

### Data

| 参数     | 说明              | 类型   | 可选值 | 默认值 |
| -------- | ----------------- | ------ | ------ | ------ |
| label | 文本信息 | ReactNode | 否 | --
| value | 选中的值 | `string`, `number` | 否 | --
| disabled | 不可用 | boolean | -- | --