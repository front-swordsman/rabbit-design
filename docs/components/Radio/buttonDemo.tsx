import * as React from 'react'
import { Radio } from 'components'

const options = [{
  label: '篮球',
  value: 'basketball'
}, {
  label: '足球',
  value: 'football'
}, {
  label: '滑板',
  value: 'skateboard'
}, {
  label: '跳伞',
  value: 'parachute'
}]

export default function RadioDemo() {
  const [value, setValue] = React.useState('basketball')
  console.log(value)
  return (
    <div style={{ padding: 20 }}>
      <h3>填充</h3>
      <Radio options={options} value={value} onChange={(value: string) => setValue(value)} shape='button' buttonStyle='solid'/>
      <h3>描边(默认)</h3>
      <Radio options={options} value={value} onChange={(value: string) => setValue(value)} shape='button' />
      <h3>大尺寸</h3>
      <Radio options={options} value={value} onChange={(value: string) => setValue(value)} shape='button' size='large'/>
      <h3>小尺寸</h3>
      <Radio options={options} value={value} onChange={(value: string) => setValue(value)} shape='button' size='small'/>
    </div>
  )
}
