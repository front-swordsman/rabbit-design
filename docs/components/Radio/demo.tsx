import * as React from 'react'
import { Radio } from 'components'

const RadioItem = Radio.item

const options = [
  {
    label: '篮球',
    value: 'basketball'
  },
  {
    label: '足球',
    value: 'football'
  },
  {
    label: '滑板',
    value: 'skateboard'
  },
  {
    label: '跳伞',
    value: 'parachute'
  }
]

export default function RadioDemo() {
  const [checked, setChecked] = React.useState(false)
  const [value, setValue] = React.useState('basketball')
  console.log(value)
  return (
    <div>
      <h3>基础用法</h3>
      <RadioItem checked={checked} onChange={checked => setChecked(checked)}>
        点击选择
      </RadioItem>
      <h3>不可用</h3>
      <RadioItem checked disabled>
        点击选择
      </RadioItem>
      <RadioItem disabled>点击选择</RadioItem>
      <h3>group 用法</h3>
      <Radio options={options} value={value} onChange={(value: string) => setValue(value)} />
    </div>
  )
}
