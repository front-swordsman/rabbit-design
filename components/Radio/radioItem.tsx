import * as React from 'react'
import cx from 'classnames'
import { RadioItemProps } from 'types/radio'
import { noop } from '../utils/tool'

const { useCallback, useImperativeHandle, useRef } = React

const defaultProps = {
  autoFocus: false,
  checked: false,
  disabled: false,
  onChange: noop,
  prefixCls: 'snake-radio-item'
}

function RadioItem(itemProps: RadioItemProps, ref: React.RefObject<{}>) {
  const props = { ...defaultProps, ...itemProps }
  const { checked, disabled, prefixCls, children, autoFocus, className, onChange } = props
  const inputRef = useRef<HTMLInputElement>()

  const getClassName = useCallback(() => {
    return cx(
      prefixCls,
      {
        [`${prefixCls}-disabled`]: disabled,
        [`${prefixCls}-checked`]: checked
      },
      className
    )
  }, [checked, disabled, className, prefixCls])

  const handleChange = useCallback(
    (e: React.ChangeEvent<HTMLInputElement>) => {
      const checked = e.target.checked
      onChange(checked)
    },
    [checked]
  )

  useImperativeHandle(ref, () => ({
    focus: () => {
      inputRef.current.focus()
    },
    blur: () => {
      inputRef.current.blur()
    }
  }))

  return (
    <label className={getClassName()}>
      <span className={`${prefixCls}-select`}>
        <input
          type="radio"
          autoFocus={autoFocus}
          checked={checked}
          disabled={disabled}
          ref={inputRef}
          className={`${prefixCls}-input`}
          onChange={handleChange}
        />
        <span className={`${prefixCls}-select-inner`} />
      </span>
      <span className={`${prefixCls}-label`}>{children}</span>
    </label>
  )
}

export default React.forwardRef(RadioItem)
