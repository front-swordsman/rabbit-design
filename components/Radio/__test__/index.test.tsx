import * as React from 'react'
import { render, fireEvent } from 'react-testing-library'
import Radio from '../index'

const RadioItem = Radio.item

describe('radio test', () => {
  const options = [
    {
      label: '篮球',
      value: 'basketball'
    },
    {
      label: '足球',
      value: 'football'
    },
    {
      label: '滑板',
      value: 'skateboard'
    },
    {
      label: '跳伞',
      value: 'parachute'
    }
  ]

  it('snapshot', () => {
    const radioItem = render(<RadioItem />)
    const checkedRadioItem = render(<RadioItem checked />)
    const disabledRadioItem = render(<RadioItem disabled />)
    const radio = render(<Radio options={options} value="basketball" />)
    const disabledRadio = render(<Radio options={options} value="" disabled />)
    const buttonRadioOutline = render(<Radio options={options} value="" shape="button" />)
    const buttonRadioSolid = render(
      <Radio options={options} value="" shape="button" buttonStyle="outline" />
    )
    expect(radioItem.container.innerHTML).toMatchSnapshot()
    expect(checkedRadioItem.container.innerHTML).toMatchSnapshot()
    expect(disabledRadioItem.container.innerHTML).toMatchSnapshot()
    expect(radio.container.innerHTML).toMatchSnapshot()
    expect(disabledRadio.container.innerHTML).toMatchSnapshot()
    expect(buttonRadioOutline.container.innerHTML).toMatchSnapshot()
    expect(buttonRadioSolid.container.innerHTML).toMatchSnapshot()
  })

  it('fire change', () => {
    function RadioDemo() {
      const [checked, setChecked] = React.useState(false)
      return (
        <RadioItem checked={checked} onChange={checked => setChecked(checked)}>
          测试
        </RadioItem>
      )
    }
    const { getByText } = render(<RadioDemo />)
    const radioNode = getByText(/测试/i).previousSibling.firstChild as HTMLInputElement
    fireEvent.click(radioNode)
    expect(radioNode.checked).toBe(true)
  })

  it('radio group', () => {
    render(<Radio options={options} value="" />)
    const radioItem = document.querySelectorAll('.snake-radio-item')
    expect(radioItem.length).toBe(4)
  })

  it('fire group change', () => {
    function Group() {
      const [value, setValue] = React.useState<string>('basketball')
      return <Radio options={options} value={value} onChange={(value: string) => setValue(value)} />
    }
    const { getByText } = render(<Group />)
    const secondNode = getByText(/足球/i).previousSibling.firstChild as HTMLInputElement
    fireEvent.click(secondNode)
    expect(secondNode.checked).toBe(true)
  })
})
