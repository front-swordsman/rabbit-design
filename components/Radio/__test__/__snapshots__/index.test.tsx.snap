// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`radio test snapshot 1`] = `"<label class=\\"snake-radio-item\\"><span class=\\"snake-radio-item-select\\"><input type=\\"radio\\" class=\\"snake-radio-item-input\\"><span class=\\"snake-radio-item-select-inner\\"></span></span><span class=\\"snake-radio-item-label\\"></span></label>"`;

exports[`radio test snapshot 2`] = `"<label class=\\"snake-radio-item snake-radio-item-checked\\"><span class=\\"snake-radio-item-select\\"><input type=\\"radio\\" class=\\"snake-radio-item-input\\" checked=\\"\\"><span class=\\"snake-radio-item-select-inner\\"></span></span><span class=\\"snake-radio-item-label\\"></span></label>"`;

exports[`radio test snapshot 3`] = `"<label class=\\"snake-radio-item snake-radio-item-disabled\\"><span class=\\"snake-radio-item-select\\"><input type=\\"radio\\" disabled=\\"\\" class=\\"snake-radio-item-input\\"><span class=\\"snake-radio-item-select-inner\\"></span></span><span class=\\"snake-radio-item-label\\"></span></label>"`;

exports[`radio test snapshot 4`] = `"<div class=\\"snake-radio snake-radio-normal snake-radio-default snake-radio-outline\\"><label class=\\"snake-radio-item snake-radio-item-checked\\"><span class=\\"snake-radio-item-select\\"><input type=\\"radio\\" class=\\"snake-radio-item-input\\" checked=\\"\\"><span class=\\"snake-radio-item-select-inner\\"></span></span><span class=\\"snake-radio-item-label\\">篮球</span></label><label class=\\"snake-radio-item\\"><span class=\\"snake-radio-item-select\\"><input type=\\"radio\\" class=\\"snake-radio-item-input\\"><span class=\\"snake-radio-item-select-inner\\"></span></span><span class=\\"snake-radio-item-label\\">足球</span></label><label class=\\"snake-radio-item\\"><span class=\\"snake-radio-item-select\\"><input type=\\"radio\\" class=\\"snake-radio-item-input\\"><span class=\\"snake-radio-item-select-inner\\"></span></span><span class=\\"snake-radio-item-label\\">滑板</span></label><label class=\\"snake-radio-item\\"><span class=\\"snake-radio-item-select\\"><input type=\\"radio\\" class=\\"snake-radio-item-input\\"><span class=\\"snake-radio-item-select-inner\\"></span></span><span class=\\"snake-radio-item-label\\">跳伞</span></label></div>"`;

exports[`radio test snapshot 5`] = `"<div class=\\"snake-radio snake-radio-normal snake-radio-disabled snake-radio-default snake-radio-outline\\"><label class=\\"snake-radio-item snake-radio-item-disabled\\"><span class=\\"snake-radio-item-select\\"><input type=\\"radio\\" disabled=\\"\\" class=\\"snake-radio-item-input\\"><span class=\\"snake-radio-item-select-inner\\"></span></span><span class=\\"snake-radio-item-label\\">篮球</span></label><label class=\\"snake-radio-item snake-radio-item-disabled\\"><span class=\\"snake-radio-item-select\\"><input type=\\"radio\\" disabled=\\"\\" class=\\"snake-radio-item-input\\"><span class=\\"snake-radio-item-select-inner\\"></span></span><span class=\\"snake-radio-item-label\\">足球</span></label><label class=\\"snake-radio-item snake-radio-item-disabled\\"><span class=\\"snake-radio-item-select\\"><input type=\\"radio\\" disabled=\\"\\" class=\\"snake-radio-item-input\\"><span class=\\"snake-radio-item-select-inner\\"></span></span><span class=\\"snake-radio-item-label\\">滑板</span></label><label class=\\"snake-radio-item snake-radio-item-disabled\\"><span class=\\"snake-radio-item-select\\"><input type=\\"radio\\" disabled=\\"\\" class=\\"snake-radio-item-input\\"><span class=\\"snake-radio-item-select-inner\\"></span></span><span class=\\"snake-radio-item-label\\">跳伞</span></label></div>"`;

exports[`radio test snapshot 6`] = `"<div class=\\"snake-radio snake-radio-normal snake-radio-button snake-radio-outline\\"><label class=\\"snake-radio-item\\"><span class=\\"snake-radio-item-select\\"><input type=\\"radio\\" class=\\"snake-radio-item-input\\"><span class=\\"snake-radio-item-select-inner\\"></span></span><span class=\\"snake-radio-item-label\\">篮球</span></label><label class=\\"snake-radio-item\\"><span class=\\"snake-radio-item-select\\"><input type=\\"radio\\" class=\\"snake-radio-item-input\\"><span class=\\"snake-radio-item-select-inner\\"></span></span><span class=\\"snake-radio-item-label\\">足球</span></label><label class=\\"snake-radio-item\\"><span class=\\"snake-radio-item-select\\"><input type=\\"radio\\" class=\\"snake-radio-item-input\\"><span class=\\"snake-radio-item-select-inner\\"></span></span><span class=\\"snake-radio-item-label\\">滑板</span></label><label class=\\"snake-radio-item\\"><span class=\\"snake-radio-item-select\\"><input type=\\"radio\\" class=\\"snake-radio-item-input\\"><span class=\\"snake-radio-item-select-inner\\"></span></span><span class=\\"snake-radio-item-label\\">跳伞</span></label></div>"`;

exports[`radio test snapshot 7`] = `"<div class=\\"snake-radio snake-radio-normal snake-radio-button snake-radio-outline\\"><label class=\\"snake-radio-item\\"><span class=\\"snake-radio-item-select\\"><input type=\\"radio\\" class=\\"snake-radio-item-input\\"><span class=\\"snake-radio-item-select-inner\\"></span></span><span class=\\"snake-radio-item-label\\">篮球</span></label><label class=\\"snake-radio-item\\"><span class=\\"snake-radio-item-select\\"><input type=\\"radio\\" class=\\"snake-radio-item-input\\"><span class=\\"snake-radio-item-select-inner\\"></span></span><span class=\\"snake-radio-item-label\\">足球</span></label><label class=\\"snake-radio-item\\"><span class=\\"snake-radio-item-select\\"><input type=\\"radio\\" class=\\"snake-radio-item-input\\"><span class=\\"snake-radio-item-select-inner\\"></span></span><span class=\\"snake-radio-item-label\\">滑板</span></label><label class=\\"snake-radio-item\\"><span class=\\"snake-radio-item-select\\"><input type=\\"radio\\" class=\\"snake-radio-item-input\\"><span class=\\"snake-radio-item-select-inner\\"></span></span><span class=\\"snake-radio-item-label\\">跳伞</span></label></div>"`;
